src/s3path.jl

```diff
@@ -322,17 +322,23 @@ function FilePathsBase.sync(f::Function, src::AbstractPath, dst::S3Path; delete=
     end
 end
 
-function Base.readdir(fp::S3Path)
+function Base.readdir(fp::S3Path; join=false)
     if isdir(fp)
         k = fp.key
         # Only list the files and "dirs" within this S3 "dir"
         objects = s3_list_objects(fp.config, fp.bucket, k; delimiter="")
 
         # Only list the basename and not the full key
-        basenames = unique!([s3_get_name(k, string(o["Key"])) for o in objects])
+        names = unique!([s3_get_name(k, string(o["Key"])) for o in objects])
 
         # Lexographically sort the results
-        return sort!(filter!(!isempty, basenames))
+        sort!(filter!(!isempty, names))
+        
+        if join
+            names = [ FilePathsBase.join(fp, name) for name in names]
+        end
+
+        return names
```

Comment:
This could probably be written more simply as:

```julia
# Only list the basename and not the full key
results = unique!([s3_get_name(k, string(o["Key"])) for o in objects])

# Filter out any empty object names which are valid in S3
filter!(!isempty, results)

# Sort results if sort=true
sort && sort!(results)

# Return results, possibly joined with the root path if join=true
return join ? joinpath.(fp, results) : results
```

Reply:
looks good!